state/application.go

```diff
@@ -640,29 +654,109 @@ func (a *Application) IsExposed() bool {
 	return a.doc.Exposed
 }
 
+// ExposedEndpoints returns a list of explicitly selected endpoints for this
+// application that should be reachable once the application is exposed. An
+// empty returned list means that all endpoints should be reachable.
+func (a *Application) ExposedEndpoints() []string {
+	if len(a.doc.ExposedEndpoints) == 0 {
+		return nil
+	}
+	return a.doc.ExposedEndpoints
+}
+
+// ExposeToSpaceIDs returns a list of space IDs that should be able to reach
+// this application's ports when the application is exposed.
+func (a *Application) ExposeToSpaceIDs() []string {
+	if len(a.doc.ExposeToSpaceIDs) == 0 {
+		return nil
+	}
+	return a.doc.ExposeToSpaceIDs
+}
+
+// ExposeToCIDRs returns a list of CIDRs that should be able to reach this
+// application's ports when the application is exposed.
+func (a *Application) ExposeToCIDRs() []string {
+	if len(a.doc.ExposeToCIDRs) == 0 {
+		return nil
+	}
+	return a.doc.ExposeToCIDRs
+}
+
 // SetExposed marks the application as exposed.
 // See ClearExposed and IsExposed.
-func (a *Application) SetExposed() error {
-	return a.setExposed(true)
+func (a *Application) SetExposed(endpoints, toSpaceIDs, toCIDRs []string) error {
+	if len(endpoints) != 0 {
+		bindings, _, err := readEndpointBindings(a.st, a.globalKey())
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		for _, endpoint := range endpoints {
+			// The default space entry has an empty string as its
+			// key but is not a valid endpoint.
+			if _, found := bindings[endpoint]; !found || endpoint == "" {
```

Comment:
Yet another thing that I missed and fixed via force-pushing. The `""` endpoint represents the default space binding but it is not a valid endpoint name ;-)

Reply:
Oof. I should have seen that.